packages/ember-glimmer/tests/integration/syntax/each-in-test.js

```diff
@@ -0,0 +1,142 @@
+import { set } from 'ember-metal/property_set';
+import { strip } from '../../utils/abstract-test-case';
+import { applyMixins } from '../../utils/abstract-test-case';
+import { moduleFor } from '../../utils/test-case';
+import {
+  BasicConditionalsTest,
+  SyntaxCondtionalTestHelpers,
+  TruthyGenerator,
+  FalsyGenerator,
+  // ObjectTestCases
+} from '../../utils/shared-conditional-tests';
+
+class EachInTest extends BasicConditionalsTest {
+
+  get truthyValue() { return { 'Not Empty': 1 }; }
+  get falsyValue() { return {}; }
+
+}
+
+applyMixins(EachInTest,
+
+  SyntaxCondtionalTestHelpers,
+
+  new TruthyGenerator([
+    // TODO: figure out what the rest of the cases are
+    { foo: 1 }
+  ]),
+
+  new FalsyGenerator([
+    // TODO: figure out what the rest of the cases are
+    {},
+    Object.create({ 'Not Empty': 1 }),
+    undefined,
+    null
+  ])
+
+  // TODO(mmun): Add support for object proxies and
+  // include the ObjectTestCases mixin.
+);
+
+moduleFor('@htmlbars Syntax test: {{#each-in}}', class extends EachInTest {
+
+  templateFor({ cond, truthy, falsy }) {
+    return `{{#each-in ${cond}}}${truthy}{{else}}${falsy}{{/each-in}}`;
+  }
+
+  [`@test it repeats the given block for each item in the hash`]() {
+    this.render(strip`
+      <ul>
+        {{#each-in categories as |category count|}}
+          <li>{{category}}: {{count}}</li>
+        {{/each-in}}
+      </ul>
+    `, {
+      categories: {
+        'Smartphones': 8203,
+        'JavaScript Frameworks': Infinity
+      }
+    });
+
+    this.assertStableRerender();
+
+    this.assertHTML(strip`
+      <ul>
+        <li>Smartphones: 8203</li>
+        <li>JavaScript Frameworks: Infinity</li>
+      </ul>
+    `);
+
+    this.runTask(() => set(this.context, 'categories', {
+      'Smartphones': 100
+    }));
```

Comment:
This is not possible in HTMLBars because

```js
set(this.context, 'categories.Smartphones', 100)
```

does not dirty anything because there is no stream observing `categories.Smartphones`. This would require a "star" observer or hooking Ember.set, but either way it would be a change from the current semantics.